lsp/zk.lua

```diff
@@ -1,74 +1,117 @@
 ---@brief
 ---
---- https://github.com/mickael-menu/zk
+--- https://github.com/zk-org/zk
 ---
 --- A plain text note-taking assistant
 
-local function find_zk_root(startpath)
-  for dir in vim.fs.parents(startpath) do
-    if vim.fn.isdirectory(vim.fs.joinpath(dir, '.zk')) == 1 then
-      return dir
+---List notes
+---@param client vim.lsp.Client
+---@param bufnr integer
+---@param opts table
+---@param action fun(path: string, title: string)
+local function zk_list(client, bufnr, opts, action)
+  opts = vim.tbl_extend('keep', { select = { 'path', 'title' } }, opts or {})
+  client:exec_cmd({
+    title = 'ZkList',
+    command = 'zk.list',
+    arguments = { vim.api.nvim_buf_get_name(bufnr), opts },
+  }, { bufnr = bufnr }, function(err, result)
+    if err ~= nil then
+      vim.api.nvim_echo({ { 'zk.list error\n' }, { vim.inspect(err) } }, true, {})
+      return
     end
-  end
+    if result == nil then
+      return
+    end
+
+    local paths = vim.tbl_map(function(item)
+      return item.path
+    end, result)
+    local titles = vim.tbl_map(function(item)
+      return item.title
+    end, result)
+    vim.ui.select(titles, {}, function(item, idx)
+      if item ~= nil then
+        action(paths[idx], item)
```

Comment:
`tbl_map` uses `pairs()` which is not guaranteed to have stable order, even on the same data. Could do this instead, which avoids needing tbl_map at all:

```lua
vim.ui.select(result, {
  format_item = function(item)
    return item.title
  end,
}, function(item)
  if item ~= nil then
    action(item.path, item.title)
  end
end)
```